File: new/PalindromeLinkedList.java
package leetcode231_240;

import java.util.ArrayList;

/**Given a singly linked list, determine if it is a palindrome.
 Follow up: Could you do it in O(n) time and O(1) space?
 * Created by eugene on 16/4/29.
 */
public class PalindromeLinkedList {

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    //TODO 此法空间O(1),但是很容易出错. 先找到中点并同时逆转前半部分,再从中点往头/尾遍历并比较.
    public boolean isPalindrome(ListNode head) {
        if(head == null) return true;
        ListNode p1 = head;
        ListNode p2 = head;
        ListNode p3 = p1.next;
        ListNode pre = p1;
        //find mid pointer, and reverse head half part
        while(p2.next != null && p2.next.next != null) {
            p2 = p2.next.next;
            pre = p1;
            p1 = p3;
            p3 = p3.next;
            p1.next = pre;
        }
        //odd number of elements, need left move p1 one step
        if(p2.next == null) {
            p1 = p1.next;
        }
        //compare from mid to head/tail
        while(p3 != null) {
            if(p1.val != p3.val) {
                return false;
            }
            p1 = p1.next;
            p3 = p3.next;
        }
        return true;
    }


    public boolean isPalindrome1(ListNode head) {
        ArrayList<Integer> list = new ArrayList<>();
        ListNode p = head;
        while (p!=null){
            list.add(p.val);
            p = p.next;
        }
        int len = list.size();
        if (len<=1) return true;
        for (int i=0; i<len/2; i++){
            if (list.get(i).compareTo(list.get(len-i-1))!=0) return false;
        }
        return true;
    }

}

File: Python/PalindromeLinkedList.cpp
package leetcode231_240;

import java.util.ArrayList;

/**Given a singly linked list, determine if it is a palindrome.
 Follow up: Could you do it in O(n) time and O(1) space?
 * Created by eugene on 16/4/29.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head) return true;
        ListNode* slow = head, * fast = head;
        while (fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        slow = reverse(slow);
        while (slow && slow->val == head->val){
            slow = slow->next;
            head = head->next;
        }
        return slow == nullptr;
    }
    ListNode* reverse(ListNode* head){
        if(!head || !head->next) return head;
        ListNode* second = head->next;
        ListNode* nHead = reverse(second);
        second->next = head;
        head->next = nullptr;
        return nHead;
    }
};
//public class PalindromeLinkedList {
//
//    public static class ListNode {
//        int val;
//        ListNode next;
//        ListNode(int x) { val = x; }
//    }
//
//    /**图 https://leetcode.com/discuss/78152/java-easy-to-understand
//     * 先找到中点,逆转后半部分,比较
//     * @param head
//     * @return
//     */
//    public boolean isPalindrome(ListNode head) {
//        ListNode fast = head;
//        ListNode slow = head;
//        while(fast != null && fast.next != null) {
//            slow = slow.next;
//            fast = fast.next.next;
//        }
//        slow = reverseList(slow);
//        while(slow != null && head.val == slow.val) {
//            head = head.next;
//            slow = slow.next;
//        }
//        return slow == null;
//    }
//    private ListNode reverseList(ListNode head) {
//        if(head==null || head.next==null) return head;
//        ListNode second = head.next;    //pin the second node
//        ListNode rest = reverseList(second);
//        second.next = head;
//        head.next = null;   //TODO set first's next to be null
//        return rest;
//    }
//
//}
